### PR TITLE
Add Learnable Query

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request

## Description

This adds a Learnable Query constructor that creates a query with some randomness to possibly be able to "ensemble" predictions through querying the Perceiver output multiple times.

Fixes issue #16 

## How Has This Been Tested?

Unit tests

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: perceiver_pytorch/queries.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+        Args:
+            channel_dim: Channel dimension for the output of the network
+            query_shape: The final shape of the query, generally, the (T, H, W) of the output
+        """
+        super(LearnableQuery, self).__init__()
+        self.query_shape = query_shape
+        # Need to get Fourier Features once and then just append to the output
+        self.fourier_features = encode_position(
+            1,
```
Comment (peterdudfield) [inline perceiver_pytorch/queries.py:39]: do you mind adding what this first variable is?

File: perceiver_pytorch/queries.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            axis=query_shape,
+            max_frequency=max_frequency,
+            frequency_base=frequency_base,
+            num_frequency_bands=num_frequency_bands,
+            sine_only=sine_only,
+        )
+        _LOG.debug(self.fourier_features.shape)
```
Comment (peterdudfield) [inline perceiver_pytorch/queries.py:46]: is this needed?

File: tests/test_queries.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        query_shape=(6, 16, 16),
+        conv_layer=layer_shape,
+        max_frequency=64.0,
+        frequency_base=2.0,
+        num_frequency_bands=128,
+        sine_only=False,
+    )
+    x = torch.randn((4, 6, 12, 16, 16))
+    out = query_creator(x)
+    assert out.shape == (4, 1536, 803)
```
Comment (peterdudfield) [inline tests/test_queries.py:21]: Are these dimensions correct? Can you add a comment showing why these are right?

File: perceiver_pytorch/queries.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        sine_only: bool = False,
+    ):
+        """
+        Learnable Query with some inbuilt randomness to help with ensembling
+
+        Args:
+            channel_dim: Channel dimension for the output of the network
+            query_shape: The final shape of the query, generally, the (T, H, W) of the output
+        """
+        super(LearnableQuery, self).__init__()
```
Comment (peterdudfield) [inline perceiver_pytorch/queries.py:35]: What does this do? Sorry - I feel like I should know this

File: tests/test_queries.py (hunk 3 above)
Comment (jacobbieker) [reply tests/test_queries.py]: Yes, and yeah, just did

File: perceiver_pytorch/queries.py (hunk 2 above)
Comment (jacobbieker) [reply perceiver_pytorch/queries.py]: Probably not,  removed!

File: perceiver_pytorch/queries.py (hunk 4 above)
Comment (jacobbieker) [reply perceiver_pytorch/queries.py:35]: I'm not entirely sure, but it is how you are supposed to subclass the `torch.nn.Module` [Doc](https://pytorch.org/docs/stable/generated/torch.nn.Module.html)

File: tests/test_queries.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        conv_layer=layer_shape,
+        max_frequency=64.0,
+        frequency_base=2.0,
+        num_frequency_bands=128,
+        sine_only=False,
+    )
+    x = torch.randn((4, 6, 12, 16, 16))
+    out = query_creator(x)
+    # Output is flattened, so should be [B, T*H*W, C]
+    assert out.shape == (4, 16 * 16 * 6, 803)
```
Comment (peterdudfield) [inline tests/test_queries.py:22]: Sorry, I probably just dont understand, where does the 803 come from?

Comment (jacobbieker) [reply tests/test_queries.py:22]: That is the number of channels of the `channel_dim` so 32 + the number of Fourier Features, which for this is 771

Comment (jacobbieker) [reply tests/test_queries.py:22]: And the number of Fourier Features comes from (num input axis)*((num_freq_bands*2)+1) so for this 3*(128*2+1) = 771

Comment (peterdudfield) [reply tests/test_queries.py:24]: Thanks for adding this

**Review by peterdudfield — APPROVED** [review]
Looks good to me, 

had for me to judge if this is the right method - is it copied from somewhere? or reference in a  paper?

**Comment by jacobbieker** [comment]
> Looks good to me,
> 
> had for me to judge if this is the right method - is it copied from somewhere? or reference in a paper?

This isn't copied from anywhere, just the general idea from using a random latent space to generate predictions from GANs, somewhat the Skillful Nowcasting GAN paper I guess

File: perceiver_pytorch/queries.py (hunk 4 above)
Comment (JackKelly) [reply perceiver_pytorch/queries.py:35]: In Python 3 you can shorten this to:

```suggestion
        super().__init__()
```

I've submitted a [pull request](https://github.com/pytorch/pytorch/pull/65748) to change the PyTorch documentation :)

File: perceiver_pytorch/queries.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+from perceiver_pytorch.utils import encode_position
+from math import prod
+import einops
+import logging
+
+_LOG = logging.getLogger("perceiver.queries")
+_LOG.setLevel(logging.WARN)
+
+
+class LearnableQuery(torch.nn.Module):
```
Comment (JackKelly) [inline perceiver_pytorch/queries.py:13]: On the topic of position encoding for the queries...

I think the position encoding needs to be 'internally consistent' across each entire example.  That is, the position encoding should be consistent across data inputs _and_ the queries so the model can see that, say, the last timestep of the recent history in the input is immediately before the first timestep of the query.  And that, in general, the model can see that the timeseries of recent history and the timeseries of queries are two parts of a contiguous timeseries.

To give a concrete example: If the recent history in the input spans 11:00 to 11:55, and the query is for 12:00 to 12:55, then we want the model to see that the first timestep of the query is 5 minutes after the last timestep of the input... if that makes sense?!

One way to do this might be to encode the positions once for all timesteps in the example (i.e. the concatenation of the recent history and the forecast timesteps), and then concatenate the last `forecast_timesteps` of the position encoding to the queries?  Or something like that?!  Not sure what's best!

Perhaps this could wait for a future PR though!

This is just a hunch, of course!

And, ultimately, we might want the position encoding to include both the relative position ("the i<sup>th</sup> element in the array") and the absolute position in time and space ("2pm in South London").  But we can worry about that later :)

File: perceiver_pytorch/queries.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        num_frequency_bands: int = 64,
+        frequency_base: float = 2.0,
+        sine_only: bool = False,
+    ):
+        """
+        Learnable Query with some inbuilt randomness to help with ensembling
+
+        Args:
+            channel_dim: Channel dimension for the output of the network
+            query_shape: The final shape of the query, generally, the (T, H, W) of the output
```
Comment (JackKelly) [inline perceiver_pytorch/queries.py:33]: Sorry for being slow but what's the `T` in `(T, H, W)`?  Time?

File: perceiver_pytorch/queries.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        )
+        self.channel_dim = channel_dim
+        if (
+            conv_layer == "3d" and len(self.query_shape) == 3
+        ):  # If Query shape is for an image, then 3D conv won't work
+            conv = torch.nn.Conv3d
+        elif conv_layer == "2d":
+            conv = torch.nn.Conv2d
+        else:
+            raise ValueError(f"Value for 'layer' is {conv_layer} which is not one of '3d', '2d'")
```
Comment (JackKelly) [inline perceiver_pytorch/queries.py:54]: Nice!

**Review by JackKelly — APPROVED** [review]
Looks really good!  I do love this idea of using random queries to create an ensemble!

File: perceiver_pytorch/queries.py (hunk 7 above)
Comment (jacobbieker) [reply perceiver_pytorch/queries.py:33]: Yes, its the number of timesteps

File: perceiver_pytorch/queries.py (hunk 6 above)
Comment (jacobbieker) [reply perceiver_pytorch/queries.py:13]: Yeah, I might add that in a follow up PR, but it should be fairly easy to add those as options. I think it's a good idea to try at least!
